Add threaded websocket client

Numerlor opened the pull request:
Adds a thread safe websocket client that can dispatch messages to multiple subscribed consumers through queues. Example usage here https://github.com/Momentous-Muses/code-jam/blob/fd7eb93abb625390e2342296e3ae22639d88e949/client/websocket_client/client.py#L115-L134

New message classes are created as Message subclasses, e.g.
```py
@dataclass
class CustomMessage(websocket_client.Message):
    channel = "test_channel"
    type = "test_type"

    @classmethod
    def from_json_dict(cls, json_):
        return cls_created_from_json

    def to_json(self) -> str:
        return json.dumps(as_dict(self))
```
after the Message subclass is created, when a message with the given type is received on a channel, the class is created from the received json and sent to all queues on the given channel.

The uassage I have in mind is that the tkinter client will create the dispatcher, keep it around and for each queue it subscribes for (each channel will represent a different, separate part of the communication), it'll poll it with a function scheduled with `app.after`

Commit:
Add threaded websocket client

ooliver1 commented on client/websocket_client/__init__.py:
```diff
@@ -0,0 +1,5 @@
+from .messages import Message
+from .utils import UNINITIALIZED, ResultEvent
+
+__all__ = ["Message", "ResultEvent", "UNINITIALIZED", "WebsocketMessageDispatcher"]
```
```suggestion
__all__ = ("Message", "ResultEvent", "UNINITIALIZED", "WebsocketMessageDispatcher")
```
`__all__` does not need mutation

ooliver1 reviewed (changes requested):
Otherwise, looks good!

Commit:
rename channel to domain

Commit:
Add PySide6

Commit:
serialize to json dict instead of json directly

this avoids unnecessary json loads in every method
and allows the sender to attach additional data

Commit:
Add util for creating asyncio tasks that log their errors

Commit:
use error logging create_task

Commit:
set __feature__ import as first party for isort

Commit:
use unique channels for communication

Commit:
send the client id with all messages

Commit:
use logging interpolation instead of f-string

Commit:
use same attribute name as json key

Commit:
provide defaults for Messages serialization/deserialization methods

Commit:
update example usage

Commit:
mark generated uuid as possibly none

Numerlor committed:
use tuple for __all__ instead of list

Co-authored-by: Oliver Wilkes <oliverwilkes2006@icloud.com>

Commit:
Do not allow non-ready connections to attempt scheduling sends

Commit:
Handle refused channel connection requests

c-gaskell commented:
There's merge conflicts for poetry.lock - i'm not entirely sure how this file works, does anyone else know how this conflict should be resolved?

Commit:
Merge remote-tracking branch 'origin/main' into threaded-websocket-client

Numerlor commented:
for lock file conflicts, it just has to be generated again; I've resolved it and brought the branch up to date with main

c-gaskell commented:
Just waiting for the review from @ooliver1 before I can merge as he requested changes, once that's done I'm good to merge
